### DISCUSSION TITLE
Duda para modelar con Strings

### DISCUSSION THREAD

**Discussion by francorojo** [dds-jv/foro#170] [Consultas de objetos] [answered]
Buenas, otra vez xd, tengo una duda de cuando es buena justificacion usar Strings para modelar en el diseño. En este caso traigo el parcial donde surgio mi duda: 
Apuestas Mundialistas -> https://docs.google.com/document/d/1_I8GNONbn0jo782OxLNflu467XdLa9Ry7ocUh_AErew/edit#
![image](https://user-images.githubusercontent.com/49093389/89738598-63832780-da50-11ea-8e12-ef513eb1bec6.png)
Mi duda seria, la variable goles en la clase Resultado modelada como una relacion entre el equipo (que es un string) y los jugadores que hicieron los goles (donde sus nombres son strings). Entiendo que otra opcion tal vez no es buena, porque enums no tienen mucho sentido ya que el dominino puede ser super extenso, y crear clases equipo y jugadores no tiene sentido ahora en este contexto porque tal vez no harian nada y serian clases sin comportamiento.
Peeero me surge la inquietud de si hace falta justificar que tal vez que el ingreso de los datos lo va a hacer el cliente que conoce el dominio, tal vez que esos datos van a estar normalizados o tal vez a traves de la interfaz se eligira el boton que represente al jugador pero no va a ingresar el nombre a mano directamente el cliente. 
Por que no puede haber el problema que vimos en clase de que por ejemplo se ingrese [Messi, Mesi] y por un error de tipeo son dos jugadores distintos, o incluso ingresar equipos que no existen? 
Una idea que me puede surgir es crear un repo o una clase con nombre "Torneo" (que seria un repo pero capaz el nombre es mas representativo) que contenga a los equipos y sus jugadores respectivos en el dominio.

(Puede ser util porque cada torneo tiene sus equipos y se podria hacer algo del estilo:
_torneoPrimeraDivision = new Torneo() 
torneoSegundaDivision = new Torneo()_
y agregar o sacar equipos y jugadores luego depende los cambios del torneo por ejemplo:
_torneoPrimeraDivision.agregar("PepitaPasion", ["Pepe", "Lala"]))
torneoPrimeraDivision.nuevoJugador("PepitaPasion", "el mago")_
)
Entonces cada vez que agregamos un partido se puede validar que el equipo y sus jugadores existan realmente en el sistema y sino informar que se ha introducido un nombre erroneo.

- Para concluir, perdon que se hizo largo, si pienso la solucion como pase la foto del diagrama hace falta alguna justificacion para bancar los strings o estaria bien asi sin aclarar nada? (no digo que no este aclarado en la solucion del drive, digo si yo tengo q justificar cuando comunique la solucion en otra ocasión similar) Y si pienso la segunda solucion, esta demas esa clase Torneo que banque la persistencia? (Por tal vez se puede hacer algo con bases de datos pero todavia no tengo idea)

**Muchas gracias!!!**

**Comment by gastonprieto** [comment]
Buenas @francorojo !

Uff, sacaste uno de ultratumba jajaja Dejo el [enunciado original](https://docs.google.com/document/d/1We4eDWfwYXEL54rrcXYWMDnkcIMztx2jU7L_dmRVU1E/edit?usp=sharing) 

> Peeero me surge la inquietud de si hace falta justificar que tal vez que el ingreso de los datos lo va a hacer el cliente que conoce el dominio, tal vez que esos datos van a estar normalizados o tal vez a traves de la interfaz se eligira el boton que represente al jugador pero no va a ingresar el nombre a mano directamente el cliente.
Por que no puede haber el problema que vimos en clase de que por ejemplo se ingrese [Messi, Mesi] y por un error de tipeo son dos jugadores distintos, o incluso ingresar equipos que no existen?

Si, si fueras por clases podrías explicar esto que comentas por ejemplo. Incluso lo mas probable es que quieras modelarlo así para que los objetos que pueden ser elegidos para las apuestas ya esten definidos. 

Entonces **si estaría bueno** que comentes porque tenes esa data class, que si bien tal vez es un code smell, es algo que tendría sentido tenerlo.

> Entonces cada vez que agregamos un partido se puede validar que el equipo y sus jugadores existan realmente en el sistema y sino informar que se ha introducido un nombre erroneo.

Suena rara la responsabilidad de ese objeto. Entiendo que solo estarías cambiando de lugar ese estado como Strings, en vez de gnerarle una abstracción a esto que agrupa `n` jugadores, que sería el equipo.

> Para concluir, perdon que se hizo largo, si pienso la solucion como pase la foto del diagrama hace falta alguna justificacion para bancar los strings o estaria bien asi sin aclarar nada? (no digo que no este aclarado en la solucion del drive, digo si yo tengo q justificar cuando comunique la solucion en otra ocasión similar) 

Lo ideal sería que si no tenes otras decisiones que creas mas importantes, y si estuviste un rato pensando cual elegir, entonces estaría bueno que lo cuentes.

> Y si pienso la segunda solucion, esta demas esa clase Torneo que banque la persistencia? (Por tal vez se puede hacer algo con bases de datos pero todavia no tengo idea)

Para mi torneos no esta tan mal, pero si me molesta que siga laburando a nivel String y no objetos que representen algo en el dominio. Tal vez `Torneo` podría conocer `Equipo`s y sirva para saber si un equipo por el que apuestan existe y luego a ese equipo preguntarle si tiene cierto jugador. En vez de seguir manejando los Strings y ver como relacionarlos (parece un primitive obsession)

**Comment by francorojo** [comment, marked as answer]
Muchas graciaaas!, clarisimo, voy a tratar de justificar todo por las dudas si me decido por una opcion de entre varias.
Saludooos!
